macaroonbakery/bakery.py

```diff
@@ -0,0 +1,280 @@
+# Copyright 2016 Canonical Ltd.
+# Licensed under the LGPLv3, see LICENCE file for details.
+
+import base64
+import json
+import webbrowser
+
+import macaroons
+import requests
+
+
+def deserialize(json_macaroon):
+    '''Deserialize a JSON macaroon into a macaroon object.
+
+    @param the JSON macaroon to deserialize as a dict.
+    @return the deserialized macaroon object.
+    '''
+    return macaroons.deserialize(to_base64(json_macaroon))
+
+
+def serialize_macaroon_string(macaroon):
+    '''Serialize macaroon object to string.
+
+    @param macaroon object to be serialized.
+    @return a string serialization form of the macaroon.
+    '''
+    return base64.urlsafe_b64decode(
+        _add_padding(macaroon.serialize_json()))
+
+
+def to_base64(json_macaroon):
+    '''Base 64 encode a JSON macaroon.
+
+    @param the JSON macaroon as a dict to encode.
+    @return the encoded macaroon.
+    '''
+    return base64.urlsafe_b64encode(json.dumps(json_macaroon))
+
+
+def _add_padding(s):
+    '''Add padding to base64 encoded string
+    libmacaroons does not give padded base64 string from serialization.
+
+    @param string s to be padded.
+    @return a padded string.
+    '''
+    return s + '=' * ((4 - (len(s) % 4)) % 4)
+
+
+class DischargeError(Exception):
+    '''Exception that is been raised by Bakery public methods when a discharge
+       error happens.
+    '''
+    pass
+
+
+class Bakery(object):
+    ''' Bakery is a wrapper around requests that ensures
+        that macaroons headers are sent to the server if present or gather the
+        discharged macaroons automatically before sending the request.
+    '''
+    def __init__(self, initial_macaroons=None, discharge_token=None,
+                 timeout=30):
+        '''
+        @param initial_macaroons initial macaroon to be used in headers as a
```

Comment:
It seems a bit odd to me that the initial macaroons are in external (base64-encoded) format rather than being an array of actual macaroon objects. I think it might be better to use
one representation consistently in the API, and that should probably be macaroon
objects. I'm open to arguments to the contrary though.